L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTKMTFProducer.cc

```diff
@@ -0,0 +1,185 @@
+
+  // ----------member data ---------------------------
+  edm::EDGetTokenT<l1t::MuonStubCollection> stubToken_;
+  KMTF* kmtf_;
+  unsigned int Nprompt;
+  unsigned int Ndisplaced;
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+Phase2L1TGMTKMTFProducer::Phase2L1TGMTKMTFProducer(const edm::ParameterSet& iConfig)
+    : stubToken_(consumes<l1t::MuonStubCollection>(iConfig.getParameter<edm::InputTag>("stubs"))),
+      kmtf_(new KMTF(iConfig.getParameter<int>("verbose"), iConfig.getParameter<edm::ParameterSet>("algo"))),
+      Nprompt(iConfig.getParameter<uint>("Nprompt")),
+      Ndisplaced(iConfig.getParameter<uint>("Ndisplaced")) {
+  produces<std::vector<l1t::SAMuon> >("prompt").setBranchAlias("prompt");
+  produces<std::vector<l1t::SAMuon> >("displaced").setBranchAlias("displaced");
+  produces<std::vector<l1t::KMTFTrack> >("kmtfTracks");
+}
+
+Phase2L1TGMTKMTFProducer::~Phase2L1TGMTKMTFProducer() {}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void Phase2L1TGMTKMTFProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+  edm::Handle<l1t::MuonStubCollection> stubHandle;
+  iEvent.getByToken(stubToken_, stubHandle);
+
+  l1t::MuonStubRefVector stubs;
+  for (uint i = 0; i < stubHandle->size(); ++i) {
+    l1t::MuonStubRef stub(stubHandle, i);
+    if (stub->bxNum() == 0)
+      stubs.push_back(stub);
+  }
+
+  // KMTF
+  std::vector<SAMuon> prompt;
+  std::vector<SAMuon> displaced;
+  std::pair<std::vector<l1t::KMTFTrack>, std::vector<l1t::KMTFTrack> > kmtfOutput = kmtf_->process(stubs, 0, 32);
+  std::vector<l1t::KMTFTrack> kmtfTracks;
+  for (const auto& track : kmtfOutput.first) {
+    kmtfTracks.push_back(track);
+    l1t::SAMuon p(track.p4(),
+                  (track.curvatureAtVertex() < 0),
+                  track.ptPrompt(),
+                  track.coarseEta(),
+                  track.phiAtMuon() / (1 << 5),
+                  0,
+                  0,
+                  track.stubs().size() - 1);
+    p.setTF(l1t::tftype::bmtf);
+    int bstart = 0;
+    wordtype word(0);
+    bstart = wordconcat<wordtype>(word, bstart, 1, 1);
+    bstart = wordconcat<wordtype>(word, bstart, p.hwCharge(), 1);
+    bstart = wordconcat<wordtype>(word, bstart, p.hwPt(), BITSPT);
+    bstart = wordconcat<wordtype>(word, bstart, p.hwPhi(), BITSPHI);
+    bstart = wordconcat<wordtype>(word, bstart, p.hwEta(), BITSETA);
+    bstart = wordconcat<wordtype>(word, bstart, p.hwD0(), BITSSAD0);
+    bstart = wordconcat<wordtype>(word, bstart, track.rankPrompt(), 8);
+
+    for (const auto& stub : track.stubs())
+      p.addStub(stub);
+    p.setWord(word);
+    prompt.push_back(p);
+  }
+
+  for (const auto& track : kmtfOutput.second) {
+    kmtfTracks.push_back(track);
+    ap_int<7> dxy = track.dxy() * ap_ufixed<8, 1>(1.606);
+    l1t::SAMuon p(track.p4(),
```

Comment:
@zhenbinwu I just realized I have  a bug here on l132 track.p4() has to be replaced by track.displacedP4() . Can we add it ? If not we can do another PR later

Reply:
Nice catch. Ok, fixed.